File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    
    static int max;
    static ArrayList<Integer>[] graph;
    static Map<Integer, Integer> dist;
    static boolean[] visit;
    
    public int solution(int n, int[][] edge) {
        int answer = 0;
        graph = new ArrayList[n + 1];
        dist = new HashMap<>();
        visit = new boolean[n + 1];
        
        for(int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<Integer>();
        }
        
        for(int i = 0; i < edge.length; i++) {
            int x = edge[i][0];
            int y = edge[i][1];
            
            graph[x].add(y);
            graph[y].add(x);
        }
        
        bfs(1);
        
        for(int x : dist.keySet()) {
            if(dist.get(x) == max) answer++;
        }
        
        return answer;
    }
    
    static void bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        max = Integer.MIN_VALUE;
        
        q.add(start);
        dist.put(start, 0);
        visit[start] = true;
        
        while(!q.isEmpty()) {
            int x = q.poll();
            
            for(int y : graph[x]) {
                if(visit[y]) continue;
                dist.put(y, dist.get(x) + 1);
                visit[y] = true;
                if(dist.get(y) >= max) max = dist.get(y);
                q.add(y);
            }
        }
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
class Queue {
    constructor() {
        this.queue = [];
    }
    
    enqueue(n) {
        this.queue.push(n);
    }
    
    dequeue() {
        return this.queue.shift();
    }
    
    isEmpty() {
        return this.queue.length === 0;
    }
}

function solution(n, edge) {
    for(let i = 0; i <= n; i++) {
        adj[i] = [];
        dist[i] = 20000;
    }
    
    for(let vertex of edge) {
        const [a, b] = vertex;
        
        adj[a].push(b);
        adj[b].push(a);
    }
    
    for(let i = 1; i <= n; i++) {
        adj[i].sort();
    }
    
    bfs();
    
    let cnt = 0;
    let max = 0;
    for(let i = 2; i <= n; i++) {
        if(max < dist[i]) {
            cnt = 1;
            max = dist[i];
            continue;
        }
        if(max === dist[i]) {
            cnt++;
            continue;
        }
    }
    
    return cnt;
}

let adj = [];
let dist = [];

const bfs = () => {
    const q = new Queue();
    q.enqueue(1);
    dist[1] = 0;
    
    while(!q.isEmpty()) {
        const x = q.dequeue();
        
        for(let y of adj[x]) {
            if(dist[y] <= dist[x] + 1) continue;
            if(y === 1) continue;
            
            dist[y] = dist[x] + 1;
            q.enqueue(y);
        }
    }
}